server/src/main/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregator.java

```diff
@@ -40,22 +40,23 @@
 /**
  * An aggregator that computes approximate counts of unique values.
  */
-public class CardinalityAggregator extends NumericMetricsAggregator.SingleValue {
+public abstract class CardinalityAggregator extends NumericMetricsAggregator.SingleValue {
 
-    private final int precision;
+    protected final int precision;
+    protected ValuesSourceConfig valuesSourceConfig;
     private final ValuesSource valuesSource;
 
     // Expensive to initialize, so we only initialize it when we have an actual value source
     @Nullable
-    private HyperLogLogPlusPlus counts;
+    protected HyperLogLogPlusPlus counts;
```

Comment:
Is this still nullable?

Reply:
Yes, because `createUnmapped` still makes an instance of `CardinalityAggregator` with a null values source.  I might get around to cleaning that up at some point (I'd like to) but I didn't do it in this pass.

Reply:
ah! But could it be final?